qutip/solver/dysolve_propagator.py

```diff
@@ -0,0 +1,399 @@
+from qutip import Qobj, qeye_like
+import numpy as np
+import itertools
+from scipy.special import factorial
+
+__all__ = ['DysolvePropagator', 'dysolve_propagator']
+
+
+class DysolvePropagator:
+    """
+    A generator of propagator using the Dysolve algorithm
+    (see https://arxiv.org/abs/2012.09282).
+
+    Parameters
+    ----------
+    H_0 : Qobj
+        The hamiltonian of the system. Must be diagonal.
+
+    X : Qobj
+        A cosine perturbation applied on the system.
+
+    omega : float
+        The frequency of the cosine perturbation.
+
+    options : dict, optional
+        Extra parameters when creating a DysolvePropagator instance.
+        "max_order" is a given integer to indicate the highest order
+        of approximation used to compute the propagators (default is 4).
+        "a_tol" is the absolute tolerance used when computing the propagators
+        (default is 1e-10).
+
+    Notes
+    -----
+    The system's hamiltonian must be of the form
+    H = H_0 + cos(omega*t)X for Dysolve to work.
+
+    """
+
+    def __init__(
+        self,
+        H_0: Qobj,
+        X: Qobj,
+        omega: float,
+        options: dict[str] = None,
+    ):
+        # System
+        self.H_0 = H_0
+        self.eigenenergies = H_0.diag()
+        self.X = X.transform(H_0)
```

Comment:
This is not how transform work...
The input should be a basis: all ket in of the new basis.
If `H_0` is diagonal, it should be `X.transform(qeye_like(H_0))`.

Usage:
```
        self.eigenenergies, self.basis = H_0.eigenstates()
        self.X = X.transform(self.basis)
```